### PR TITLE
[feature/captcha_v2] Refactor captcha classes and implement turnstile captcha

### PR DESCRIPTION
**Pull request by marc1706**
Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket:

https://tracker.phpbb.com/browse/PHPBB-17413
https://tracker.phpbb.com/browse/PHPBB-17414
https://tracker.phpbb.com/browse/PHPBB-17415

### CODE REVIEW

File: phpBB/phpbb/captcha/plugins/turnstile.php [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+			// Return without checking against server without a turnstile response
+			return false;
+		}
+
+		// Retrieve form data for verification
+		$form_data = [
+			'secret'			=> $this->config['captcha_turnstile_secret'],
+			'response'			=> $turnstile_response,
+			'remoteip'			=> $this->request->header('CF-Connecting-IP'),
+			//'idempotency_key'	=> $this->confirm_id, // check if we need this
```
Comment (Derky) [inline phpBB/phpbb/captcha/plugins/turnstile.php:143]: Remaining todo, please remove if not necessary.

**Review by CHItA — APPROVED** [review]

File: phpBB/phpbb/captcha/plugins/turnstile.php (hunk 1 above)
Comment (CHItA) [reply phpBB/phpbb/captcha/plugins/turnstile.php:143]: i don't think we want to allow multiple validations.

File: phpBB/phpbb/captcha/plugins/turnstile.php [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		{
+			// Return without checking against server without a turnstile response
+			return false;
+		}
+
+		// Retrieve form data for verification
+		$form_data = [
+			'secret'			=> $this->config['captcha_turnstile_secret'],
+			'response'			=> $turnstile_response,
+			'remoteip'			=> $this->request->header('CF-Connecting-IP'),
```
Comment (CHItA) [inline phpBB/phpbb/captcha/plugins/turnstile.php:142]: Is turnstile only works if the site is proxied via cloudflare? If not, probably we should do our own ip magic here.

Comment (CHItA) [reply phpBB/phpbb/captcha/plugins/turnstile.php:142]: e.g. recaptcha uses $user->ip

Comment (marc1706) [reply phpBB/phpbb/captcha/plugins/turnstile.php:142]: Good point, will go the recaptcha route

File: phpBB/phpbb/captcha/plugins/turnstile.php (hunk 1 above)
Comment (marc1706) [reply phpBB/phpbb/captcha/plugins/turnstile.php:143]: Yeah, I agree that this shouldn't be needed and will remove it.

**Review by CHItA — APPROVED** [review]
lgtm

**Pull request merged**